Task Inline Edit

aydreeihn opened the pull request:
Allow inline edit for the following on tasks:
- custom fields
- Status
- Department
- Due Date
- Assign

aydreeihn committed:
Task Inline Edit

Allow inline edit for the following on tasks:
- custom fields
- Status
- Department
- Due Date
- Assign

aydreeihn committed:
Task Inline Edit

- Put updateField in VerySimpleModel so that we don't have to have redundant code in the Ticket and Task classes.
- Change Task filter from 'Due Soon' to 'Due Date'
- Make sure we add any new custom fields on the task view

JediKev commented on include/staff/templates/task-view.tmpl.php:
```diff
@@ -403,34 +410,73 @@ class="faded">&mdash; '.__('None').' &mdash;</span>'; ?></td>
+                                if ($isFile && !$isEmpty) {
+                                    echo sprintf('<span id="field_%s" %s >%s</span><br>', $id,
+                                        $class,
+                                        $v ?: '<span class="faded">&mdash;' . __('Empty') .  '&mdash; </span>');
```
`Format::strip_tags()`

aydreeihn committed:
Inline Edit Long Answer

This commit ensures that if a long answer field is cleared using inline edit, we display the faded '--Empty--' string so that the agent still has the option to edit the value again.

aydreeihn commented:
@JediKev I have rebased and added the commit to ensure that we are able to clear long answer fields and edit them again. This is ready to be merged now.

aydreeihn committed:
Inline Edit Checkboxes

This commit adds the extra switch statement we needed for BooleanFields so that checkboxes will always display as 'Yes' or 'No' rather than 0 or 1.

aydreeihn committed:
Oops: Lint Fixes

Fix undefined variable error.

aydreeihn committed:
Issue: updateField

We need to have the updateField function in the Ticket and Task class rather than in VerySimpleModel to avoid having an irrelevant function in the ORM.

aydreeihn commented:
@protich @JediKev I ended up putting the updateField function in the task and ticket class because Threadable is an interface and I couldn't create a new class that extended VerySimpleModel because I got the error 'Model does not define meta.table'.

protich commented on include/class.orm.php:
```diff
@@ -755,6 +755,7 @@ function copy() {
         $dup->__new__ = true;
         return $dup;
     }
+
```
drop this errant space / changes to this file.

protich commented on include/ajax.tasks.php:
```diff
@@ -211,6 +211,259 @@ function edit($tid) {
+                $state = 'open';
+                break;
+            case 'close':
+                if (!$role->hasPerm(Ticket::PERM_CLOSE))
```
You mean `Task::PERM_CLOSE` ?

protich commented:
I'm also a little confused why you're using Ticket Statuses! Task doesn't have status_id -  it uses flag field to set open / closed bit.

aydreeihn committed:
Clean Up Task Status

This commit cleans up the code used to change the status of a Task.
- Remove unneeded code for inline task status edit
- Fix mass process verbs
- Check to see if a status goes in as a string or object for setStatus
- Make sure we're looking at Task Permissions
- Make sure task status can only be default open or closed

aydreeihn commented on include/ajax.tasks.php:
```diff
@@ -655,7 +616,7 @@ function massProcess($action, $w=null)  {
             $info['status'] = $info['status'] ?: 'closed';
-            $perm = $action = '';
+            $perm = '';
             switch ($info['status']) {
```
We need to make sure we don't clear $action here because we use it to get the correct verb for the success message.

aydreeihn commented on include/class.task.php:
```diff
@@ -172,16 +172,6 @@ function getCloseDate() {
         return $this->isClosed() ? $this->closed : '';
     }
 
-    function getStatusId() {
```
I decided this wasn't a good way to do things, so I got rid of it altogether. Can't rely on the status names staying the same...

aydreeihn commented on include/staff/templates/status-options.tmpl.php:
```diff
@@ -26,11 +26,14 @@
+    $state = ($object->getStatus() == 'Open') ? 'open' : 'closed';
+    foreach (TicketStatusList::getStatuses(
+                array('states' => $states)) as $status) {
+        if (isset($actions[$status->getState()])
```
This will make sure we only show Open or Closed in the dropdown list for tasks since those are the only options you can switch to.

aydreeihn commented:
@protich Ok, did some cleaning up on this. Sorry about that.

aydreeihn committed:
Redo Task Status Edit

Inline edit for task status doesn't need to have the dropdown option since we can only set the status to Open or Closed.

protich commented on include/class.task.php:
```diff
@@ -545,7 +576,7 @@ function setStatus($status, $comments='', &$errors=array()) {
 
         $ecb = null;
-        switch($status) {
+        switch(is_string($status) ? $status : strtolower($status->getName())) {
```
do you need to revert this change?

aydreeihn replied:
Yes, reverted now.
